src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+
+        protected override SyntaxNode AddLinqUsing(SyntaxNode root)
+        {
+            const string linqNamespaceName = "System.Linq";
+            if (root is CompilationUnitSyntax compilationUnit)
+            {
+                var existingUsings = compilationUnit.Usings;
+                if (!existingUsings.Any(existingUsing => existingUsing.Name.ToString().Equals(linqNamespaceName)))
+                {
+                    var linqName = SyntaxFactory.ParseName(linqNamespaceName);
+                    return compilationUnit.AddUsings(SyntaxFactory.UsingDirective(linqName));
+                }
+            }
+
+            return root;
+        }
+
+        private static bool IsList(ITypeSymbol typeSymbol, SemanticModel semanticModel)
+            => Equals(typeSymbol.OriginalDefinition, semanticModel.Compilation.GetTypeByMetadataName(typeof(List<>).FullName));
+
+        private abstract class AbstractConverter
+        {
+            protected static QueryExpressionSyntax CreateQueryExpression(ForEachStatementSyntax forEachStatement, ImmutableArray<SyntaxNode> convertingNodes, ExpressionSyntax selectExpression)
+                => SyntaxFactory.QueryExpression(
+                    CreateFromClause(forEachStatement),
+                    SyntaxFactory.QueryBody(
+                        SyntaxFactory.List(convertingNodes.Select(qc => CreateQueryClause(qc).WithoutTrivia())),
+                        // The current coverage of foreach statements to support does not need to use query continuations.
+                        SyntaxFactory.SelectClause(selectExpression).WithoutTrivia(), continuation: null).WithoutTrivia()).WithAdditionalAnnotations(Formatter.Annotation);
+
+            private static QueryClauseSyntax CreateQueryClause(SyntaxNode node)
+            {
+                switch (node.Kind())
+                {
+                    case SyntaxKind.VariableDeclarator:
+                        var variable = (VariableDeclaratorSyntax)node;
+                        return SyntaxFactory.LetClause(variable.Identifier, variable.Initializer.Value);
+                    case SyntaxKind.ForEachStatement:
+                        return CreateFromClause((ForEachStatementSyntax)node);
+                    case SyntaxKind.IfStatement:
+                        return SyntaxFactory.WhereClause(((IfStatementSyntax)node).Condition);
+                }
+
+                throw new ArgumentException(node.Kind().ToString());
+            }
+
+            private static FromClauseSyntax CreateFromClause(ForEachStatementSyntax forEachStatement)
+                => SyntaxFactory.FromClause(
+                    forEachStatement.Type is IdentifierNameSyntax identifierName && identifierName.Identifier.ValueText == "var" ? null : forEachStatement.Type,
+                    forEachStatement.Identifier,
+                    forEachStatement.Expression);
+        }
+
+        private sealed class YieldReturnConverter : AbstractConverter, IConverter
```

Comment:
why does this need to implement IConverter?  Why doesn't AbstractConverter already implement that interface?